journal/src/main/java/io/camunda/zeebe/journal/file/JournalSegment.java

```diff
@@ -154,6 +159,9 @@ private MappedJournalSegmentWriter createWriter(final long lastWrittenIndex) {
    */
   void onReaderClosed(final MappedJournalSegmentReader reader) {
     readers.remove(reader);
+    if (markedForDeletion && readers.isEmpty()) {
```

Comment:
:clown_face: This is executed in the reader thread, right? I would prefer if we keep the concept of a strong writer somehow (including delete operations), such that this is then the job of the writer thread. But I'm not sure how to do this without introducing the concept of thread context into the journal :sweat: It should normally be fast, but still...

Reply:
:x: Additionally, if this is called on the reader thread, then there could be a race condition since `markedForDeletion` is not volatile nor atomic, so multiple threads may try to mark the segment for deletion?

Reply:
:face_with_head_bandage: You are right about the race condition. Multiple reader threads can try to delete the file at the same time. It is also possible that the readers doesn't read the latest value of `markedForDeletion` since it is not volatile, and thus the segment would never get deleted.

Reply:
> clown_face This is executed in the reader thread, right? I would prefer if we keep the concept of a strong writer somehow (including delete operations), such that this is then the job of the writer thread. But I'm not sure how to do this without introducing the concept of thread context into the journal sweat It should normally be fast, but still...

It would be difficult to do this, since the journal doesn't have threads. The semantic deletion is still done by the writer - executed in the raft thread. Only deferred deletion is executed in the reader which is closed last. 

We can think about introducing thread context in to the journal. It might be good because currently we are assuming that there is only one writer writing to the journal. We enforce it outside of the journal by ensuring that only raft thread is writing. But there is no way to verify this with in the journal.

Reply:
Regarding the last comment, OTOH I'm not sure what's the best approach for this. Extend `SingleThreadContext` to add something like `#runInContext(Runnable)`, which either executes inline if you're already in the thread, or enqueues the runnable otherwise? I'd have to think about it and alternatives a bit first.

Reply:
Let's add this to the list in #6221 We can discuss what would be a good approach.

Reply:
Can you do this? Thanks!

Reply:
:heavy_check_mark: